### ISSUE TITLE
has_internet() doesn't work properly if a proxy is utilized

### ISSUE THREAD

**Issue by lshep** [jeroen/curl#153]
Our package AnnotationHub utilizes the curl::has_internet() to check for an internet connection.  However it seems it always returns FALSE when a proxy is specified despite having internet connection.

https://github.com/Bioconductor/AnnotationHub/issues/4

Is there some way to account for proxy use in the has_internet function?

**Comment by jeroen** [comment]
Hmm interesting. So that means `curl::nslookup()` doesn't work at all for you?

**Comment by zymla** [comment]
For me, behind a corporate firewall, `curl::nslookup()` is not working.
According to https://blogs.msdn.microsoft.com/oldnewthing/20150916-00/?p=91581 , `getAddrInfo()` will not go through the proxy as it is not an HTTP request (and there is just no way to pass a handle with proxy paramters to it anyway).
Maybe having a parameter in `has_internet()` that switch to an alternate method to `nslookup()` when a handle with proxy parameters is provided would be a good option?

**Comment by jeroen** [comment]
But what would that alternate method be? Shouldn't `getAddrInfo()` just work if your network/proxy is correctly configured?

**Comment by zymla** [comment]
1/ Something along:
```
has_internet <- function(handle = NULL){
  if(is.null(handle){
    !is.null(nslookup("google.com", error = FALSE))
  } else {
    !inherits(try(curl_fetch_memory('https://www.google.com', handle = handle)), "try-error")
  }
}

h <- 
  curl::new_handle(
    proxy = curl::ie_get_proxy_for_url('https://www.google.com'),
    proxyusername = 'username',
    proxypassword = 'qwerty'
  )

has_internet(handle = h)
```
2/ It is quite common for corporate proxies to block DNS lookups, as it is done by the HTTP proxy. 
The proxy gets the host name in the HTTP request from the client, looks it up, and and with the IP address it opens the TCP connection to the server. The client doesn't need to know the IP address of the server (just that of the proxy). Usually in a corporate setting, the end user has no way to bypass the configuration of his machine nor that of the proxy :(

**Comment by bedantaguru** [comment]
> For me, behind a corporate firewall, `curl::nslookup()` is not working.
> According to https://blogs.msdn.microsoft.com/oldnewthing/20150916-00/?p=91581 , `getAddrInfo()` will not go through the proxy as it is not an HTTP request (and there is just no way to pass a handle with proxy paramters to it anyway).
> Maybe having a parameter in `has_internet()` that switch to an alternate method to `nslookup()` when a handle with proxy parameters is provided would be a good option?

Same situation here. Also, I feel there should be a vignette or some documentation in the help page of `curl_options {curl}` which tells the user how to configure a proxy. Still, I'm at the dark about it.

**Comment by jeroen** [comment]
Can you show your `curl::ie_proxy_info()` output?

**Comment by bedantaguru** [comment]
Here is the output with masked IP.

``` r
curl::ie_proxy_info()
#> $AutoDetect
#> [1] FALSE
#>
#> $AutoConfigUrl
#> [1] "http://172.xx.xx.xx:8443/proxypac.pac"
#>
#> $Proxy
#> [1] "172.xx.xx.xx:8080"
#>
#> $ProxyBypass
#> [1] "10.*;172.*;mail.rbi.org.in"
```

<sup>Created on 2019-08-22 by the [reprex package](
https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by bedantaguru** [comment]
Here is the result with IP masked 

``` r
curl::ie_proxy_info() 
#> $AutoDetect
#> [1] FALSE
#> 
#> $AutoConfigUrl
#> [1] "http://172.35.10.87:8543/proxypac.pac"
#> 
#> $Proxy
#> [1] "172.35.10.87:3228"
#> 
#> $ProxyBypass
#> [1] "10.*;172.*;mail.rbi.org.in;autodiscover.rbi.org.in;ev.rbi.org.in;ecms.rbi.org.in;login.rbi.org.in;ekp.rbi.org.in;ekplogin.rbi.org.in;ekpportal.rbi.org.in;ekpsite.rbi.org.in;dmis.rbi.org.in;ieslogin.rbi.org.in;ieshelp.rbi.org.in;mumco.rbi.org.in;libco.rbi.org.in;cdbms.rbi.org.in;hrms.rbi.org.in;hrmsco.rbi.org.in;secweb.rbi.org.in;helpdesk.rbi.org.in;iccomssws.rbi.org.in;iccomsweblogictest.rbi.org.in;cosmos.rbi.org.in;hriserver.rbi.org.in;cas.rbi.org.in;cpad.rbi.org.in;dbodreturns.rbi.org.in;ims.r"
```

<sup>Created on 2019-08-22 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by jeroen** [comment]
OK and what do you see for `curl::ie_get_proxy_for_url("https://www.google.com")`

**Comment by jeroen** [comment]
@bedantaguru can you confirm if `curl::ie_get_proxy_for_url("https://www.google.com")` gives the correct result in your case?

**Comment by bedantaguru** [comment]
@jeroen terribly sorry for the long wait. For a few days, there was a change in our corporate network proxy, which unknowingly blocked GitHub. So I was accessing Github from home only. Currently, they unblocked it. 

Again apologies for the long delay. Here is a few info which may be of your assistance.

``` r

# on corporate network
curl::ie_proxy_info() 
#> $AutoDetect
#> [1] FALSE
#> 
#> $AutoConfigUrl
#> [1] "http://172.35.10.87:8543/proxypac.pac"
#> 
#> $Proxy
#> [1] "172.35.10.87:3228"
#> 
#> $ProxyBypass
#> [1] "10.*;172.*;mail.rbi.org.in;autodiscover.rbi.org.in;ev.rbi.org.in;ecms.rbi.org.in;login.rbi.org.in;ekp.rbi.org.in;ekplogin.rbi.org.in;ekpportal.rbi.org.in;ekpsite.rbi.org.in;dmis.rbi.org.in;ieslogin.rbi.org.in;ieshelp.rbi.org.in;mumco.rbi.org.in;libco.rbi.org.in;cdbms.rbi.org.in;hrms.rbi.org.in;hrmsco.rbi.org.in;secweb.rbi.org.in;helpdesk.rbi.org.in;iccomssws.rbi.org.in;iccomsweblogictest.rbi.org.in;cosmos.rbi.org.in;hriserver.rbi.org.in;cas.rbi.org.in;cpad.rbi.org.in;dbodreturns.rbi.org.in;ims.r"
curl::ie_get_proxy_for_url("https://www.google.com")
#> [1] "172.35.10.87:3228"
curl::has_internet()
#> [1] FALSE

# further drill
curl::nslookup("google.com", error = FALSE)
#> NULL
.Call(curl:::R_nslookup, "google.com", FALSE)
#> NULL
```

<sup>Created on 2019-08-29 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
**Note: IP masked in similar manner. In both the cases**

Now If I unplug the corporate network and connect my mobile network (I did it with admin permission here just to test this), then I get the following output of same code. 

``` r
# same machine but on mobile (unrestricted and direct connect) network
# No proxy required in this network
curl::ie_proxy_info() 
#> $AutoDetect
#> [1] FALSE
#> 
#> $AutoConfigUrl
#> [1] "http://172.35.10.87:8543/proxypac.pac"
#> 
#> $Proxy
#> [1] "172.35.10.87:3228"
#> 
#> $ProxyBypass
#> [1] "10.*;172.*;mail.rbi.org.in;autodiscover.rbi.org.in;ev.rbi.org.in;ecms.rbi.org.in;login.rbi.org.in;ekp.rbi.org.in;ekplogin.rbi.org.in;ekpportal.rbi.org.in;ekpsite.rbi.org.in;dmis.rbi.org.in;ieslogin.rbi.org.in;ieshelp.rbi.org.in;mumco.rbi.org.in;libco.rbi.org.in;cdbms.rbi.org.in;hrms.rbi.org.in;hrmsco.rbi.org.in;secweb.rbi.org.in;helpdesk.rbi.org.in;iccomssws.rbi.org.in;iccomsweblogictest.rbi.org.in;cosmos.rbi.org.in;hriserver.rbi.org.in;cas.rbi.org.in;cpad.rbi.org.in;dbodreturns.rbi.org.in;ims.r"
curl::ie_get_proxy_for_url("https://www.google.com")
#> [1] "172.35.10.87:3228"
curl::has_internet()
#> [1] TRUE

# further drill
curl::nslookup("google.com", error = FALSE)
#> [1] "172.217.166.174"
.Call(curl:::R_nslookup, "google.com", FALSE)
#> [1] "172.217.166.174"
```

<sup>Created on 2019-08-29 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

So that means 
```r
.Call(curl:::R_nslookup, "google.com", FALSE)
```
Is not seeing the proxy. 

Please let me know if anything else is required.

**Comment by jeroen** [comment]
I have now updated `curl::has_internet()` to try to connect via a proxy server if one is detected.

**Comment by bedantaguru** [comment]
Let me try again with this update.

**Comment by bedantaguru** [comment]
Wow! it is working now 
``` r

utils::install.packages("tidycells", method = "wininet", repos = "http://cran.rstudio.com/")
#> package 'tidycells' successfully unpacked and MD5 sums checked
#> 
#> The downloaded binary packages are in
#>  C:\Users\Dell\AppData\Local\Temp\RtmpYBUDk4\downloaded_packages

curl::has_internet()
#> [1] TRUE

Sys.setenv(http_proxy="http://172.30.1.72:3128")
Sys.setenv(HTTPS_PROXY = "http://172.30.1.72:3128")

curl::has_internet()
#> [1] TRUE
```

<sup>Created on 2019-09-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>


Also, I get the following messages 
`Testing for internet connectivity via IE proxy...`
and 
`Testing for internet connectivity via https_proxy...`

thanks a lot @jeroen 👍

**Comment by bedantaguru** [comment]
Wow! it is working now 
``` r

utils::install.packages("tidycells", method = "wininet", repos = "http://cran.rstudio.com/")
#> package 'tidycells' successfully unpacked and MD5 sums checked
#> 
#> The downloaded binary packages are in
#>  C:\Users\Dell\AppData\Local\Temp\RtmpYBUDk4\downloaded_packages

curl::has_internet()
#> [1] TRUE

Sys.setenv(http_proxy="http://172.35.10.87:3228")
Sys.setenv(HTTPS_PROXY = "http://172.35.10.87:3228")

curl::has_internet()
#> [1] TRUE
```

<sup>Created on 2019-09-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>


Also, I get the following messages 
`Testing for internet connectivity via IE proxy...`
and 
`Testing for internet connectivity via https_proxy...`

thanks a lot @jeroen 👍
